js/pro/independentreserve.js

```diff
@@ -0,0 +1,284 @@
+        }, market);
+    }
+
+    async watchOrderBook (symbol, limit = undefined, params = {}) {
+        /**
+         * @method
+         * @name independentreserve#watchOrderBook
+         * @description watches information on open orders with bid (buy) and ask (sell) prices, volumes and other data
+         * @param {string} symbol unified symbol of the market to fetch the order book for
+         * @param {int|undefined} limit the maximum amount of order book entries to return
+         * @param {object} params extra parameters specific to the independentreserve api endpoint
+         * @returns {object} A dictionary of [order book structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-book-structure} indexed by market symbols
+         */
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        symbol = market['symbol'];
+        if (limit === undefined) {
+            limit = 100;
+        }
+        limit = this.numberToString (limit);
+        const url = this.urls['api']['ws'] + '/orderbook/' + limit + '?subscribe=' + market['base'] + '-' + market['quote'];
+        const messageHash = 'orderbook:' + symbol + ':' + limit;
+        const orderbook = await this.watch (url, messageHash, undefined, messageHash);
+        return orderbook.limit ();
+    }
+
+    handleOrderBook (client, message) {
+        //
+        //    {
+        //        Channel: "orderbook/1/eth/aud",
+        //        Data: {
+        //          Bids: [
+        //            {
+        //              Price: 2198.09,
+        //              Volume: 0.16143952,
+        //            },
+        //          ],
+        //          Offers: [
+        //            {
+        //              Price: 2201.25,
+        //              Volume: 15,
+        //            },
+        //          ],
+        //          Crc32: 1519697650,
+        //        },
+        //        Time: 1676150558254,
+        //        Event: "OrderBookSnapshot",
+        //    }
+        //
+        const event = this.safeString (message, 'Event');
+        const channel = this.safeString (message, 'Channel');
+        const parts = channel.split ('/');
+        const depth = this.safeString (parts, 1);
+        const baseId = this.safeString (parts, 2);
+        const quoteId = this.safeString (parts, 3);
+        const base = this.safeCurrencyCode (baseId);
+        const quote = this.safeCurrencyCode (quoteId);
+        const symbol = base + '/' + quote;
+        const orderBook = this.safeValue (message, 'Data', {});
+        const messageHash = 'orderbook:' + symbol + ':' + depth;
+        const timestamp = this.safeInteger (message, 'Time');
+        let storedOrderBook = this.safeValue (this.orderbooks, symbol);
+        if (storedOrderBook === undefined) {
+            storedOrderBook = this.orderBook ({});
+            this.orderbooks[symbol] = storedOrderBook;
+        }
+        if (event === 'OrderBookSnapshot') {
```

Comment:
@pcriadoperez sometimes deltas arrive sooner than the snapshot but we can't return the deltas right away, we need to wait until for the orderbook otherwise we return something like this: (1 ask, 0 bids, 1 ask, 1 bid, etc)

```
 p independentreserve watchOrderBook "BTC/USD"
Python v3.10.9
CCXT v2.7.84
independentreserve.watchOrderBook(BTC/USD)
{'asks': [[21609.56, 0.00012225]],
 'bids': [],
 'datetime': '2023-02-13T16:46:35.889Z',
 'nonce': None,
 'symbol': None,
 'timestamp': 1676306795889}
{'asks': [[21609.56, 0.00012225], [27905.4, 0.007]],
 'bids': [],
 'datetime': '2023-02-13T16:46:35.898Z',
 'nonce': None,
 'symbol': None,
 'timestamp': 1676306795898}
{'asks': [[21609.56, 0.00012225], [27905.4, 0.007], [27915.59, 0.1]],
 'bids': [],
 'datetime': '2023-02-13T16:46:36.026Z',
 'nonce': None,
 'symbol': None,
 'timestamp': 1676306796026}
```

Reply:
interesting, the snapshot is supposed to always arrive first. I added a check for it.